### PR TITLE
Implement norm_op by CUDA instead of Eigen.

### LINKED ISSUE #13235: fluid.layers.l2_normalize output nan when using gpu mode
fluid.layers.l2_normalize work well when using cpu mode, but it output nan changing to gpu mode

when i replace "fluid.layers.l2_normalize" by 
`    weight_norm = fluid.layers.sqrt(fluid.layers.reduce_sum(fluid.layers.square(weight), dim=1))
    weight = fluid.layers.elementwise_div(weight, weight_norm, axis=0)`,
it alse work well.

```
import os
import sys
import math
import time
import argparse
import functools
import numpy as np
import paddle
import paddle.fluid as fluid


def arc_margin_product(input, label, out_dim, m=0.2, s=20.0):
    input = fluid.layers.l2_normalize(input, axis=1)
    weight = fluid.layers.create_parameter(
                shape=[out_dim, input.shape[1]],
                dtype='float32',
                name='weight_norm',
                attr=fluid.param_attr.ParamAttr(
                              initializer=fluid.initializer.Xavier()))
    weight = fluid.layers.l2_normalize(weight, axis=1)
    #weight_norm = fluid.layers.sqrt(fluid.layers.reduce_sum(fluid.layers.square(weight), dim=1))
    #weight = fluid.layers.elementwise_div(weight, weight_norm, axis=0)
    weight = fluid.layers.transpose(weight, perm = [1, 0])
    cosine = fluid.layers.mul(input, weight)
    return cosine


def convolutional_neural_network(img, label):
    conv_pool_1 = fluid.nets.simple_img_conv_pool(
        input=img,
        filter_size=5,
        num_filters=20,
        pool_size=2,
        pool_stride=2,
        act="relu")
    conv_pool_1 = fluid.layers.batch_norm(conv_pool_1)
    conv_pool_2 = fluid.nets.simple_img_conv_pool(
        input=conv_pool_1,
        filter_size=5,
        num_filters=50,
        pool_size=2,
        pool_stride=2,
        act="relu")
    fc_1 = fluid.layers.fc(input=conv_pool_2, size=2048)
    prediction = arc_margin_product(fc_1, label, 2048)
    return prediction

def train():
    image = fluid.layers.data(name='image', shape=[1, 28, 28], dtype='float32')
    label = fluid.layers.data(name='label', shape=[1], dtype='int64')

    # model definition
    out = convolutional_neural_network(image, label)

    test_program = fluid.default_main_program().clone(for_test=True)


    #place = fluid.CUDAPlace(0)
    place = fluid.CPUPlace()
    exe = fluid.Executor(place)
    exe.run(fluid.default_startup_program())

    fetch_list_train = [out.name]

    data = np.random.uniform(0, 1, (10, 1, 28, 28)).astype(np.float32)
    label = np.arange(10).astype(np.int64).reshape(10, 1)
    loss = exe.run(test_program, fetch_list=fetch_list_train, feed={'image': data, 'label':label})
    print loss

def main():
    train()

if __name__ == '__main__':
    main()

```

### PR DESCRIPTION
**Pull request by qingqing01**
resolve #13235 #13200 

Use the network in #13235, run 10 times:
norm_op speed: 17.8719 ms ->   0.508928 ms

### CODE REVIEW

File: paddle/fluid/operators/norm_op.cu [hunk 1]
```diff
@@ -11,14 +11,151 @@ distributed under the License is distributed on an "AS IS" BASIS,
...
+  for (int i = blockIdx.x; i < num; i += gridDim.x) {
+    int base = (i / post) * post * axis_n + (i % post);
+
+    T sum = 0.0;
+    __shared__ T norm;
+    for (int j = threadIdx.x; j < axis_n; j += blockDim.x) {
+      const T x_ij = x[base + j * post];
+      sum += x_ij * x_ij;
+    }
+    T reduce_result = BlockReduce(temp_storage).Sum(sum);
```
Comment (chengduoZH) [inline paddle/fluid/operators/norm_op.cu:42]: Do you analyze whether to use `double` is better here? Because the `layer_norm_op ` ever uses `T` as the type of `reduce_result` and `sum`, but this leads to poor convergence of some models, and this [PR](https://github.com/PaddlePaddle/Paddle/pull/13232/files) fixes it.

Comment (qingqing01) [reply paddle/fluid/operators/norm_op.cu:42]: @chengduoZH I don't know why it affect convergence in some model.  But this is a simply reduce sum. 即使不实用cub，通常也是用float类型求和吧。

**Review by chengduoZH — APPROVED** [review]
